Support payment processing via platform checkout

reykjalin opened the pull request:
#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

To enable payment processing via platform checkout we need to make some changes to the payment processing, mostly in the form of filters.

Related PRs are listed below via GitHub mentions. You must use the code in the related PRs for the changes in this PR to be effective.

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Follow the testing instruction in PR 280 in the platform repo (linked below).

-------------------

- [ ] Added changelog entries to both `readme.txt` and `changelog.txt` (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : _Add link here / 'QA Testing Not Applicable'_

Commit:
Add info necessary to pay via platform checkout

reykjalin committed:
Merge branch 'develop' into add/platform-checkout-payment-processing

Commit:
Add filters necessary to pay via platform checkout

Commit:
Fix how blog_url is passed to merchant store

Commit:
add public getter for the WCPay customer service

Commit:
Send account ID with platform init request

Commit:
Handle update_order when payment method is "null"

There seem to be some cases where the payment method id is passed in as
the string "null" instead of the value null. Pretty weird, but easy to
handle.

Commit:
Add method to clone a payment method

Note: this requires checking out a corresponding PR on the server.

Commit:
Add filter for blog ID used for API calls

Commit:
Add filter for account ID used to load Stripe JS

reykjalin commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2184,6 +2185,9 @@ public function update_order_status() {
 			}
 
 			$payment_method_id = isset( $_POST['payment_method_id'] ) ? wc_clean( wp_unslash( $_POST['payment_method_id'] ) ) : '';
+			if ( 'null' === $payment_method_id ) {
+				$payment_method_id = '';
+			}
```
This was a weird little bug I ran into and I'm not exactly sure how it happens. But for whatever reason the value of `$_POST['payment_method_id']` was set to `"null"` (i.e. a string) when it was supposed to be `null`.

Honestly not sure how to reproduce, but adding the extra if-statement here fixed the issue.

reykjalin commented on includes/class-wc-payments.php:
```diff
@@ -789,16 +800,29 @@ public static function maybe_register_platform_checkout_hooks() {
 	public static function ajax_init_platform_checkout() {
 		$session_cookie_name = apply_filters( 'woocommerce_cookie', 'wp_woocommerce_session_' . COOKIEHASH );
 
+		$user        = wp_get_current_user();
+		$customer_id = self::$customer_service->get_customer_id_by_user_id( $user->ID );
+		if ( null === $customer_id ) {
+			// create customer.
+			$customer_data = WC_Payments_Customer_Service::map_customer_data( null, new WC_Customer( $user->ID ) );
+			self::$customer_service->create_customer_for_user( $user, $customer_data );
+		}
```
This approach of creating the Customer like this before seems like the simplest way to get what we want:

- A customer ID we can pass to the platform checkout.
- A customer ID associated with the customer's current session — even if they're checking out as a guest.

Some alternative approaches that might be worth looking into;

- A re-usable function we can use throughout WCPay to initialize a customer.
- Creating the customer on the platform checkout side and just pass `null` here if the customer ID hasn't been created yet.

The current implementation work, so these are mostly notes that would be best to resolve and discuss in a separate issue.

reykjalin committed:
Merge branch 'develop' into add/platform-checkout-payment-processing

reykjalin committed:
Merge branch 'develop' into add/platform-checkout-payment-processing

reykjalin committed:
Merge branch 'develop' into add/platform-checkout-payment-processing

Commit:
Add handling for payment via platform checkout

reykjalin committed:
Merge branch 'develop' into add/platform-checkout-payment-processing

reykjalin committed:
Merge branch 'develop' into add/platform-checkout-payment-processing

reykjalin committed:
Merge branch 'develop' into add/platform-checkout-payment-processing

rafaelzaleski commented:
Hey @reykjalin, I know this PR is still a draft, but I just wanted to point out that I cannot save the settings on this branch. If you are aware of that, please ignore this comment.

<img width="606" alt="Screen Shot 2022-02-04 at 10 28 12" src="https://user-images.githubusercontent.com/4153905/152537123-e44dba43-f3ba-4890-aac3-ede6c65d7291.png">

reykjalin committed:
Merge branch 'develop' into add/platform-checkout-payment-processing

reykjalin commented:
@rafaelzaleski Thanks for looking into that! I'm not so sure this is caused by this PR; it may have been a problem in `develop` the last time I updated the branch.

I updated the branch just now and hopefully that will have fixed the issue.

In any case, I'm not going to spend time on debugging the issue right now because we want to try to make changes on the server side that will make it so that we won't need the changes in this PR. If we're not successful in completely eliminating the need for the changes in this PR I'll come back to this issue 🙂

reykjalin committed:
Merge branch 'develop' into add/platform-checkout-payment-processing

reykjalin commented:
Closed in favor of https://github.com/Automattic/woocommerce-payments/pull/3811
